src/passes/Inlining.cpp

```diff
@@ -332,6 +357,432 @@ doInlining(Module* module, Function* into, const InliningAction& action) {
+
+    // The inlineable function should only have the ifs, which will call the
+    // outlined heavy work.
+    for (Index i = 0; i < numIfs; i++) {
+      // For each if, create an outlined function with the body of that if,
+      // and call that from the if.
+      auto* inlineableIf = getIf(split.inlineable->body, i);
+      auto* outlined = copyFunction(func, "outlined-B");
+      outlined->body = inlineableIf->ifTrue;
+
+      // The outlined function either returns the same results as the original
+      // one, or nothing, depending on if a value is returned here.
+      auto valueReturned =
+        func->getResults() != Type::none && outlined->body->type != Type::none;
+      outlined->setResults(valueReturned ? func->getResults() : Type::none);
+      inlineableIf->ifTrue = builder.makeCall(outlined->name,
+                                              getForwardedArgs(func, builder),
+                                              outlined->getResults());
+      if (valueReturned) {
+        inlineableIf->ifTrue = builder.makeReturn(inlineableIf->ifTrue);
+      }
+    }
+
+    // We can just leave the final value at the end, if it exists.
+
+    *inlineableOut = split.inlineable;
+    return true;
+  }
+
+  Function* copyFunction(Function* func, std::string prefix) {
+    // TODO: We copy quite a lot more than we need here, and throw stuff out.
+    //       It is simple to just copy the entire thing to get the params and
+    //       results and all that, but we could be more efficient.
+    prefix = "byn-split-" + prefix;
+    return ModuleUtils::copyFunction(
+      func,
+      *module,
+      Names::getValidFunctionName(*module, prefix + '$' + func->name.str));
+  }
+
+  // Get the i-th item in a sequence of initial items in an expression. That is,
+  // if the item is a block, it may have several such items, and otherwise there
+  // is a single item, that item itself. This basically provides a simpler
+  // interface than checking if something is a block or not when there is just
+  // one item.
+  //
+  // Returns nullptr if there is no such item.
+  static Expression* getItem(Expression* curr, Index i = 0) {
+    if (auto* block = curr->dynCast<Block>()) {
+      auto& list = block->list;
+      if (i < list.size()) {
+        return list[i];
+      }
+    }
+    if (i == 0) {
+      return curr;
+    }
+    return nullptr;
+  }
+
+  // Get the i-th if in a sequence of initial ifs in an expression. If no such
+  // if exists, returns nullptr.
+  static If* getIf(Expression* curr, Index i = 0) {
+    auto* item = getItem(curr, i);
+    if (!item) {
+      return nullptr;
+    }
+    if (auto* iff = item->dynCast<If>()) {
+      return iff;
+    }
+    return nullptr;
+  }
+
+  // Checks if an expression is very simple - something simple enough that we
+  // are willing to inline it in this optimization. This should basically take
+  // almost no cost at all to compute.
+  bool isSimple(Expression* curr) {
```

Comment:
Should the comparison instructions be simple as well?

Reply:
Do you mean the if conditions? This is called on them.

Or what do you mean here?

Reply:
I mean things like `i32.lt_u`, `i32.ge_s`, etc.

Reply:
Oh, yes, we may want those as well. For now I just added a tiny set of things, but it already handles the main things that j2cl needs. The rest is a TODO (on line 765).